### ISSUE TITLE
Consume domain events from RabbitMQ

### ISSUE THREAD

**Issue by rasmus** [eventflow/EventFlow#383]
As EventFlow is able to publish domain events to RabbitMQ, it should also be able to consume them if you want to link two bounded contexts using events.

**Comment by poumup** [comment]
Hi @rasmus could you please give me input about your vision of implementation of this issue. I would like contribute because i need this feature. Thank

**Comment by poumup** [comment]
Can someone give input about this issue ?

**Comment by rasmus** [comment]
Hi @poumup I started on the feature in PR #384, but didn't finish it yet. You could pick it up from there

**Comment by rasmus** [comment]
Ah, that PR is pretty empty. I'll see if I can give it some attention this weekend

**Comment by rasmus** [comment]
@poumup Don't I'll have a stable consumer ready this weekend. If you already have the code to consume messages from RabbitMQ in place, then EventFlow part is relatively simple.

This simple method handles the mapping of the RabbitMQ message to the domain events.
```csharp
public IDomainEvent CreateDomainEvent(RabbitMqMessage rabbitMqMessage)
{
  var metadata = new Metadata(rabbitMqMessage.Headers);
  return _eventJsonSerializer.Deserialize(
    rabbitMqMessage.Message,
     metadata);
}
```

Then you basically need to do something similar to this:
```csharp
RabbitMqMessage rabbitMqMessage = // have a look at how RabbitMqPublisher maps maps messages
var domainEvent = CreateDomainEvent(rabbitMqMessage);
await _domainEventPublisher.PublishAsync(
  new []{ domainEvent },
  cancellationToken)
  .ConfigureAwait(false);
```

I you have an example with a stable consumer that uses the "new" automatic recovery the RabbitMQ.Client recently added, then let me know.

**Comment by poumup** [comment]
Hi apologize for the delay. 
Do you want me to continue the PR on the branch: rabbitmq-domain-event-consumer ?

**Comment by poumup** [comment]
Hi,

Question.
For consume domain event, imagine 2 microservices. 
The first one publish the domain event.

The second one whant to subscribe to the event, but don't know the aggregate class and the event class.
So how to use `IDomainEvent CreateDomainEvent(RabbitMqMessage rabbitMqMessage)` ?

I suggest `.AddRabbitMqSubscriber<IIntegrationMessage>("exchangeName", "QueueName", "RoutingKey")`

where `IIntegrationMessage` is the rabbitmq serialized message

**Comment by poumup** [comment]
@rasmus what do you about my comment ?

**Comment by rasmus** [comment]
@poumup makes sense. I haven't had much "EventFlow development time" in the evenings these last few weeks though. I expect I'll pick up soon though.

**Comment by poumup** [comment]
So you have any branch ), pr or sample ?

**Comment by DureSameen** [comment]
https://github.com/DureSameen/EventFlowWithElasticSearch. I have added a repository with two branches  one WithRabbitMq and other is master. I have achieved same results in master which are also done in with rabbitMq branch.

The Idea is write at one end and read from other end.
Master Branch: 
EventFlowApi : It is write api . Configured with EventFlow , ElasticSearch, EventStore and RabbitMq. It writes to all medium through read Models.
EventFlowApi.Read: It is read api, it get ElasticSearch query handler and get the data from it.

WithRabbitMq Branch:
EventFlowApi : It is write api . Configured with EventFlow , EventStore and RabbitMq. s.
EventFlowApi.Read: It is read api, Configured with Eventflow and Elastic Search. It configured a RabbitMq subscriber at startup.which invoked Domain event subscribers when data is arrived. and insert the data in ElasticSearch.

**Comment by rasmus** [comment]
@DureSameen If you are interested you could add a link to your example from the read-me in the [external examples](https://github.com/eventflow/EventFlow#external-examples) section.

**Comment by DureSameen** [comment]
you mean you add my example in read-me external example section.. yes Please...I will be honored..

**Comment by rasmus** [comment]
@DureSameen If developers are doing awesome stuff with EventFlow, writing example applications, writing articles, then I'm all for referencing the work from the read me file. If EventFlow is to have great future with a healthy community, then I cannot by any means do everything myself or even keeping it within the EventFlow GitHub org, nor should I. EventFlow will only grow if other developers contribute as no single developer (in my opinion) has the expertise to write an enterprise grade framework by themselves.

In short, just create the PR with the addition of the link to the external section, and I'll merge it in.

If you (or others) have any ideas on how to better involve the community, then please leave your comment in #602.

**Comment by DureSameen** [comment]
I am receiving 403 error , when pushing changes in new branch.

**Comment by rasmus** [comment]
You'll need to create a pull request.

**Comment by DureSameen** [comment]
I am not allowed to push changes.

**Comment by rasmus** [comment]
No, not to the EventFlow repository, only a few members have that.

You'll need to fork the repo or do it directly in the UI. GitHub has some excellent documentation on how to do it.
https://help.github.com/en/articles/creating-a-pull-request
